docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/launcher/DockerComputerJNLPLauncher.java

```diff
@@ -0,0 +1,153 @@
+import com.github.dockerjava.api.NotFoundException;
+import com.github.dockerjava.api.command.CreateContainerCmd;
+import com.github.dockerjava.api.command.ExecCreateCmdResponse;
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.nirima.jenkins.plugins.docker.DockerComputer;
+import com.nirima.jenkins.plugins.docker.DockerTemplate;
+import com.nirima.jenkins.plugins.docker.DockerTemplateBase;
+import hudson.model.Descriptor;
+import hudson.model.TaskListener;
+import hudson.slaves.ComputerLauncher;
+import hudson.slaves.JNLPLauncher;
+import hudson.slaves.SlaveComputer;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.Charsets;
+import org.apache.commons.io.IOUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.util.Arrays;
+
+/**
+ * JNLP launcher. Doesn't require open ports on docker host.
+ *
+ * Steps:
+ * - runs container with nop command
+ * - as launch action executes jnlp connection to master
+ *
+ * @author Kanstantsin Shautsou
+ */
+public class DockerComputerJNLPLauncher extends DockerComputerLauncher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DockerComputerJNLPLauncher.class);
+
+    /**
+     * Configured from UI
+     */
+    protected JNLPLauncher jnlpLauncher;
+
+    @DataBoundConstructor
+    public DockerComputerJNLPLauncher(JNLPLauncher jnlpLauncher) {
+        this.jnlpLauncher = jnlpLauncher;
+    }
+
+    public JNLPLauncher getJnlpLauncher() {
+        return jnlpLauncher;
+    }
+
+    @Override
+    public boolean isLaunchSupported() {
+        return true;
+    }
+
+    @Override
+    public void launch(SlaveComputer computer, TaskListener listener) throws IOException, InterruptedException {
+        final PrintStream logger = listener.getLogger();
+        final DockerComputer dockerComputer = (DockerComputer) computer;
+        final String containerId = dockerComputer.getContainerId();
+        final String rootUrl = Jenkins.getInstance().getRootUrl();
+        final DockerClient connect = dockerComputer.getCloud().connect();
+
+        // exec jnlp connection in running container
+        // TODO implement PID 1 replacement
+
+        final InputStream resourceAsStream = DockerComputerJNLPLauncher.class.getResourceAsStream("init.sh");
+
+        String cdCmd = "cd /home/jenkins/";
+        String wgetSlaveCmd = "wget " + rootUrl + "jnlpJars/slave.jar -O slave.jar";
+        String jnlpConnectCmd = "java -jar slave.jar "
+                + "-jnlpUrl " + rootUrl + dockerComputer.getUrl() + "slave-agent.jnlp ";
+//                + "-secret " + dockerComputer.getJnlpMac();
+
+        String[] connectCmd = {
+                "bash", "-c", cdCmd + " && " + wgetSlaveCmd + " && " + jnlpConnectCmd
```

Comment:
Perhaps this pseudo example will help me explain:

```
cat > Dockerfile << EOF
FROM myregistry/java:8

RUN mkdir /home/jenkins
WORKDIR /home/jenkins
ADD http://myartifactory/repo/slave.jar /home/jenkins

ENTRYPOINT [ "java", "-jar", "slave.jar"]
EOF

docker build myslave .

docker run myslave -jnlpUrl http://jenkins.example.com:8080/some/jnlp/url -secret someSecret
```

Above avoids the need to execute the `cd /home/jenkins` and `wget` at runtime.  [ENTRYPOINT](https://docs.docker.com/reference/builder/#entrypoint) allows you to concatenate args from a `docker run` into a base command, rather than overwriting the command.  It can be a very powerful alternative to `CMD` in a Dockerfile.